File: Golang/String/SlidingWindow/LongestSubstr.go
/*Given a string s, find the length of the longest substring without duplicate characters*/
package main

import "fmt"

func LengthOfLongestSubString(s string) int {
	charSet := make(map[byte]bool)
	left, maxLen := 0, 0

	for right := 0; right < len(s); right++ {
		for charSet[s[right]] {
			delete(charSet, s[left])
			left++
		}

		charSet[s[right]] = true
		if right-left+1 > maxLen {
			maxLen = right - left + 1
		}
	}
	return maxLen
}

func main() {
	s := "abcabcbb"
	fmt.Println(LengthOfLongestSubString(s));
}

File: Java/StringQues/SlidingWindow/LongestSubstr.java
/**Given a string s, find the length of the longest substring without duplicate characters */
package Java.StringQues.SlidingWindow;

import java.util.HashSet;
import java.util.Set;

public class LongestSubstr {
    public static int LengthOfLongestSubstring(String s){
        Set<Character> set = new HashSet<>();
        int left = 0, right = 0, maxLen = 0;
        int n = s.length();

        while(right < n){
            char c = s.charAt(right);

            if(!set.contains(c)){
                set.add(c);
                maxLen = Math.max(maxLen, right - left + 1);
                right++;
            }else{
                set.remove(s.charAt(left));
                left++;
            }
        }
        return maxLen;
    }
    public static void main(String[] args) {
        String s = "abcabcbb";
        System.out.println(LongestSubstr.LengthOfLongestSubstring(s));
    }
}
